instrumentation/ratpack/ratpack-1.7/library/src/test/java/io/opentelemetry/instrumentation/ratpack/v1_7/server/RatpackServerApplicationTest.java

```diff
@@ -0,0 +1,103 @@
+    assertThat(app.getHttpClient().get("foo").getBody().getText()).isEqualTo("hi-foo");
+
+    testing.waitAndAssertTraces(
+        trace ->
+            trace.hasSpansSatisfyingExactly(
+                span ->
+                    span.hasName("GET /foo")
+                        .hasNoParent()
+                        .hasKind(SpanKind.SERVER)
+                        .hasAttributesSatisfyingExactly(
+                            equalTo(HTTP_ROUTE, "/foo"),
+                            equalTo(URL_PATH, "/foo"),
+                            equalTo(URL_SCHEME, "http"),
+                            satisfies(SERVER_PORT, v -> v.isInstanceOf(Long.class)),
+                            equalTo(SERVER_ADDRESS, "localhost"),
+                            equalTo(NETWORK_PROTOCOL_VERSION, "1.1"),
+                            equalTo(URL_QUERY, ""),
+                            equalTo(HTTP_REQUEST_METHOD, "GET"),
+                            equalTo(HTTP_RESPONSE_STATUS_CODE, 200L))));
+  }
+
+  @Test
+  void testPropagateTraceToHttpCalls() {
+    assertThat(app.getHttpClient().get("bar").getBody().getText()).isEqualTo("hi-bar");
+
+    testing.waitAndAssertTraces(
+        trace ->
+            trace.hasSpansSatisfyingExactly(
+                span ->
+                    span.hasName("GET /bar")
+                        .hasNoParent()
+                        .hasKind(SpanKind.SERVER)
+                        .hasAttributesSatisfyingExactly(
+                            equalTo(HTTP_ROUTE, "/bar"),
+                            equalTo(HTTP_REQUEST_METHOD, "GET"),
+                            equalTo(HTTP_RESPONSE_STATUS_CODE, 200L),
+                            equalTo(URL_PATH, "/bar"),
+                            equalTo(URL_SCHEME, "http"),
+                            satisfies(SERVER_PORT, v -> v.isInstanceOf(Long.class)),
+                            equalTo(SERVER_ADDRESS, "localhost"),
+                            equalTo(NETWORK_PROTOCOL_VERSION, "1.1"),
+                            equalTo(URL_QUERY, "")),
+                span ->
+                    span.hasName("GET")
+                        .hasKind(SpanKind.CLIENT)
+                        .hasParent(trace.getSpan(0))
+                        .hasAttributesSatisfyingExactly(
+                            equalTo(HTTP_ROUTE, "/other"),
+                            equalTo(URL_FULL, "http://localhost:" + app.getAppPort() + "/other"),
+                            equalTo(HTTP_REQUEST_METHOD, "GET"),
+                            equalTo(HTTP_RESPONSE_STATUS_CODE, 200L),
+                            satisfies(SERVER_PORT, v -> v.isInstanceOf(Long.class)),
+                            equalTo(SERVER_ADDRESS, "localhost"))));
+  }
+
+  @Test
+  void testIgnoreHandlersBeforeOpenTelemetryServerHandler() {
+    assertThat(app.getHttpClient().get("ignore").getBody().getText()).isEqualTo("ignored");
+    assertThat(testing.spans().stream().filter(span -> "GET /ignore".equals(span.getName())))
```

Comment:
In the groovy test we retry until this condition becomes true, here we assume that we already have the completed spans. Mayabe here we should also use awaitility to retry to ensure this test won't be flaky or rewrite to use `testing.waitAndAssertTraces` that also retries.